### DISCUSSION TITLE
New Reducer.dependency() Override

### DISCUSSION THREAD

**Discussion by jshier** [pointfreeco/swift-composable-architecture#2878] [General]
The 1.9 migration guide includes this example:

```swift
There is now a similar API for overriding dependencies on a reducer, Reducer/dependency(_:), which can be used like so:

MyFeature()
  .dependency(mockAPIClient)

The type of mockAPIClient determines how the dependency is overridden.
```

Two questions. First, how is this API similar to the previous example of the new `@Dependency(SomeClient.self)`? Second, what does "the type of mockAPIClient determines how the dependency is overridden" mean? How can overrides of dependencies differ in the first place?

**Comment by mbrandonw** [comment]
> First, how is this API similar to the previous example of the new `@Dependency(SomeClient.self)`?

The `@Dependency(SomeClient.self)` syntax is how you access a dependency by its type alone, and `.dependency(someClient)` is how you override that dependency by its type alone.

> Second, what does "the type of mockAPIClient determines how the dependency is overridden" mean? How can overrides of dependencies differ in the first place?

The original style of registering dependencies required you to provide a key path to locate the dependency. So you would do `@Dependency(\.someClient)` to resolve the dependency.

This new style, which isn't always appropriate to use, allows you to skip the key path and resolve the dependency from its type alone, using `@Dependency(SomeClient.self)`. And similarly, overriding the dependency also simplifies to `.dependency(someClient)` since the type of `someClient` is known.

**Comment by stephencelis** [comment]
> Two questions. First, how is this API similar to the previous example of the new `@Dependency(SomeClient.self)`? Second, what does "the type of mockAPIClient determines how the dependency is overridden" mean? How can overrides of dependencies differ in the first place?

It mainly provides parity with the new `@Environment`/`View.environment` APIs, where the "type" of value determines the key automatically:

```swift
@Environment(Model.self) var model
// ...
.environment(model)
```

vs.

```swift
@Dependency(Client.self) var client
// ...
.dependency(client)
```

Is there a way to make this clearer?

**Reply by jshier** [reply]
Mainly I was thrown the use of "similar API" in the first sentence, as, at a glance, they don't look similar. Perhaps something like "there is also a new API which uses the type of a dependency instance to override a `Reducer`'s use of that type" or something like that.

For the second sentence, the use of "how" made it seem like there was something fundamentally different about it but really it just used the type API as expected. This confusion may just my unfamiliarity with other versions of that API. Is there a `.dependency(\.someKey, instance)` version of the same thing?
